ocaml-lsp-server/src/dune.ml

```diff
@@ -233,43 +250,98 @@ end = struct
               let+ () = Progress.build_progress progress p in
               Some ()))
 
-  let diagnostic_loop client running dune_diagnostic_id diagnostics
-      ~include_promotions =
+  let registration_of_promotion (d : Drpc.Diagnostic.Promotion.t) =
+    let registrations =
+      let in_source = Drpc.Diagnostic.Promotion.in_source d in
+      let code_action =
+        let id = "ocamllsp-promote-action/" ^ in_source in
+        let method_ = "textDocument/codeAction" in
+        let registerOptions =
+          let documentSelector =
+            [ DocumentFilter.create ~pattern:in_source () ]
+          in
+          Lsp.Types.CodeActionRegistrationOptions.create ~documentSelector
+            ~codeActionKinds:[ CodeActionKind.Other "Promote" ]
+            ()
+          |> Lsp.Types.CodeActionRegistrationOptions.yojson_of_t
+        in
+        Registration.create ~id ~method_ ~registerOptions ()
+      in
+      [ code_action ]
+    in
+    RegistrationParams.create ~registrations
+
+  let fold_promotions promotions diagonstic ~f =
+    let promotions, requests =
+      Drpc.Diagnostic.promotion diagonstic
+      |> List.fold_left ~init:(promotions, []) ~f:(fun (ps, acc) promotion ->
+             let src = Drpc.Diagnostic.Promotion.in_source promotion in
+             match f promotion (String.Map.find promotions src) with
+             | None -> (ps, acc)
+             | Some (update, requests) ->
+               (* XXX not very efficient *)
+               (String.Map.update ps src ~f:(fun _ -> update), requests :: acc))
+    in
+    (promotions, List.flatten requests)
+
+  let add_promotions promotions diagnostic =
+    fold_promotions promotions diagnostic ~f:(fun promotion status ->
+        match status with
+        | None -> Some (Some promotion, [ registration_of_promotion promotion ])
+        | Some _ ->
+          (* TODO: ideally, we should update with the new promotion, but then we
+             need to store the actual registration ID *)
+          None)
+
+  let remove_promotions running promotions =
+    fold_promotions running promotions ~f:(fun promotion status ->
+        match status with
+        | None ->
+          Log.log ~section:"warning" (fun () ->
+              Log.msg "removing non existant promotion"
+                [ ( "promotion"
+                  , `String (Drpc.Diagnostic.Promotion.in_source promotion) )
+                ]);
```

Comment:
Quick question: how do you access these log messages? Are they only to be seen in tests?

Reply:
You must manually set `Lsp_fiber.Import.Log.level` to something that will display these messages.